cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/timeseries/FactTable.java

```diff
@@ -183,138 +187,183 @@ public void delete(FactScan scan) throws Exception {
     }
   }
 
-  /*
-   * Given the list of tag values, return sorted collection of tagValues that appear next to the given tagValues
-   * (on first position right after given ones) in any of the written facts in specified time range.
-   * @param tags - list of tags , we need to return the tags at the next position after these list of tags.
-   * @param startTs
-   * @param endTs
-   * @return Sorted collection of tags
+  /**
+   * Searches for first non-null valued tags in records that contain given list of tags and match given tag values in
+   * given time range. Returned tag values are those that are not defined in given tag values.
+   * @param allTagNames list of all tag names to be present in the record
+   * @param tagSlice tag values to filter by, {@code null} means any non-null value.
+   * @param startTs start of the time range, in seconds
+   * @param endTs end of the time range, in seconds
+   * @return {@link Set} of {@link TagValue}s
    * @throws Exception
    */
-  public Collection<TagValue> findNextAvailableTags(List<TagValue> tags, long startTs, long endTs) throws Exception {
-    //todo: push down this logic to server side to reduce RPC calls (CDAP-1421)
-    //todo: pass a limit on number of tags returned
-    //todo: go over aggregations that match the given tags as first tags in the same order.
-    byte[] startRow = codec.createStartRowKey(tags, null, startTs, true);
-    byte[] endRow = codec.createEndRowKey(tags, null, endTs, true);
-    endRow = Bytes.stopKeyForPrefix(endRow);
-    FuzzyRowFilter fuzzyRowFilter = createFuzzyRowFilter(new FactScan(startTs, endTs, null, tags),
-                                                         startRow, true);
-    Row rowResult;
-    int targetIndex = tags.size();
-    SortedSet<TagValue> result = Sets.newTreeSet(new Comparator<TagValue>() {
-      @Override
-      public int compare(TagValue t1, TagValue t2) {
-        int cmp = t1.getTagName().compareTo(t2.getTagName());
-        if (cmp != 0) {
-          return cmp;
-        }
-        if (t1.getValue() == null) {
-          if (t2.getValue() == null) {
-            return 0;
-          } else {
-            return -1;
-          }
-        }
-        if (t2.getValue() == null) {
-          return 1;
-        }
-        return t1.getValue().compareTo(t2.getValue());
+  // todo: pass a limit on number of tagValues returned
+  // todo: kinda not cool API when we expect null values in a map...
+  public Set<TagValue> findSingleTagValue(List<String> allTagNames, Map<String, String> tagSlice,
```

Comment:
may be rename "allTagNames" to "aggregateTags" ? confusing with the local variable allTags.